Привет, Николай! 

Меня зовут Ольга Макаревич и я буду проверять твою работу. Предлагаю продолжить общение на «ты». Но, если это не удобно, - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных. Ты уже проделала большую работу над проектом, но давай сделаем его еще лучше. 

Ниже ты найдешь мои комментарии - **пожалуйста, не перемещай, не изменяй и не удаляй их**. 

Увидев у тебя ошибку, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой начальник будет поступать так же, а я пытаюсь подготовить тебя именно к работе аналитиком. Но, если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку. 

Мои комментарии выглядят так:
        
<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера</b>

Таким цветом отмечены правильные решения.

</div>

<div class="alert alert-block alert-warning">
<b>Комментарий от ревьюера</b>

Таким цветом отмечены дополнения как сделать работу еще лучше.

</div>

<div class="alert alert-block alert-danger">
<b>Комментарий от ревьюера</b>

Таким цветом отмечены неудачные решения

</div>


Ты можешь оставлять свои комментарии под моими, отметь их каким-нибудь цветом или просто выдели отдельную ячейку.
Например, так:

<div class="alert alert-block alert-info">
<b>Комментарий от студента</b> <a class="tocSkip"></a>
    
Текст комментария
</div>

<div class="alert alert-block alert-success">
<b>Общий комментарий от ревьюера</b>
    
Николай, спасибо большое за присланную работу!  Извини, что вышла такая большая задержка с проверкой, я думаю, ты уже в курсе ситуации. Мы стараемся сократить время, надеюсь, скоро всё получится.
    
Проект получился классным! Спасибо за комментарии в течении всего проекта. Они очень хорошо помогают структурировать работу. 
    
<span class="text-danger">Есть несколько мелких ошибок, которые ты легко поправишь. Я написала комментарии, посмотри, пожалуйста.</span>
    
Жду обновленную версию. :) 
</div>

<div class="alert alert-block alert-success">
<b>Общий комментарий от ревьюера. V.2.</b>
    
Привет, Николай! Спасибо за быстрые правки! Всё получилось прекрасно!
    
Надеюсь,следующий проект будет быстрее! Удачи! :)
</div>

<div class="alert alert-block alert-warning">
<b>Для будущих проектов</b>
    
- лучше давать переменным осмысленные имена (cars, credits, events). df=dataframe. Представь, что у нас будет df1,df2,df3. Сложновато будет сразу понять, где нужные данные.
    
- вижу красные автоматические комментарии. Мы можем их удалить, чтобы работа была красивее. Стоит добавить в импорт библиотек:
    
            import warnings

        warnings.filterwarnings("ignore")


</div>

# Исследование надежности заемщиков.

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Описание данных:

- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

### Шаг 1. Обзор данных

#### Импорт библиотек и датасета, вызов метода .info

In [1]:
# Импорт библиотек и доп опции.
import pandas as pd
pd.options.mode.chained_assignment = None  # Отключим предупреждение

In [2]:
# Чтение датасета.
df = pd.read_csv('/datasets/data.csv')

print(df.columns)
print()
df.info()

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-block alert-danger">
<s><b>Комментарий от ревьюера</b>

Пропиши, пожалуйста, верный путь к файлу. В следующий раз я не смогу его открыть.
</div>


<div class="alert alert-block alert-warning">
<b>Комментарий от ревьюера</b>

Импорт библиотек и чтение файлов лучше проводить в разных ячейках.Если нам понадобится добавить библиотеку, то будет легче перезагрузить одну ячейку, а не весь проект. :)
</div>

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера. V.2.</b>

Спасибо!</div>    

В столбцах total_income и days_employed имеются пропуски, возможно, они взаимосвязаны. 

Названия столбцов корректны, изменять их не стоит.

In [3]:
display(df.describe())
df.head(15)

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

Скорее всего, пропуски в total_income и days_employed взаимосвязаны. Возможно это категории людей которые как правило не имеют дохода, пенсионеры или дети, видимо, в этом причина пропусков, далее мы проверим эту гипотезу.

In [4]:
print(f'Процент отсутствующих значений = {(21525-19351)/21525:,.1%}')

Процент отсутствующих значений = 10.1%


Довольно большой процент пропусков, нельзя оставлять его без внимания.

#### Столбец "children"

In [5]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Можно заметить, что некоторые строки имеют значение -1, и 20 детей соответственно, по всей видимости это ошибка, возможно вызванная повреждениями при записи. -1 изменим на +1, а 20 удалим из списка.

#### Столбец "days_employed"

In [6]:
display(df['days_employed'].describe())
print('Отрицательные значения:', df[df['days_employed']<0]['days_employed'].count())

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

Отрицательные значения: 15906


Большинство значений записаны неверно, а максимальные положительные значения трудового стажа ~1100 годам, на результаты исследования этот столбец не повлияет, потому мы не будем его учитывать.

#### Столбец "dob_years"

In [7]:
df['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Можно обнаружить 101 строку с значением 0, заменим далее их медианным значением.

#### Столбец "education" 

In [8]:
df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

Достаточно привести столбец в общий нижний регистр.

#### Столбец "education_id" 

In [9]:
df['education_id'].value_counts()

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

По всей видимости существует 5 ступеней образования, и они совпадают и столбцом "education" .

#### Столбец "family_status"

In [10]:
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

С столбцом всё в порядке.

#### Столбец "family_status_id"

In [11]:
df['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

Значения также совпадают с столбцом "family_status".

#### Столбец "gender"

In [12]:
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Присутствует 1 значение "XNA", его удаление не повлияет на структуру датасета, потому можно удалить эту строчку, либо изменить на любой гендер, женщин в 2 раза больше чем мужчин, потому это скорее-всего женщина.

#### Столбец "income_type"

In [13]:
df['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

Данные выглядят адекватно.

#### Столбец "debt"

In [14]:
df['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

Всё в порядке.

#### Столбец "total_income"

In [15]:
display(df[df['total_income'].isnull()].head(10))
df['total_income'].describe()

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

count    1.935100e+04
mean     1.674223e+05
std      1.029716e+05
min      2.066726e+04
25%      1.030532e+05
50%      1.450179e+05
75%      2.034351e+05
max      2.265604e+06
Name: total_income, dtype: float64

Судя по всему нет взаимосвязи в возрасте, либо занятости, по всей видимости пропуски связаны с ошибкой записи, этот столбец важен для нашего исследования, потому далее мы попытаемся заменить значения на медианные по столбцу.

#### Столбец "purpose"

In [16]:
df['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

Нужно объединить данные в общие категории (множество повторяющихся целей) сам столбец корректный.

#### Вывод:

- Удалить строки 20 в столбце "children". -1 изменим на +1.

- Удалить столбец "days employed".

- Столбец "dob years": заменить 0 на медианное значение.

- Столбец "education": привести в общий регистр.

- Столбец "gender": сменить гендер "XNA"

- Столбец "total income" заменить пропуски медианным значением.

- Столбец "purpose" объединить данные в категории

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера</b>

Отлично, есть первый взгляд на данные! Поехали!
</div>

### Шаг 2. Предобработка данных

#### Шаг 2.1 Заполнение пропусков

Заменим значения медианой, так как это количественные переменные, в которых присутствует большой "разброс" данных, стоит использовать медианное значение.

In [17]:
# медианное по ежемесячному доходу

median_income = df['total_income'].median()
print('Медиана по столбцу:', median_income)

Медиана по столбцу: 145017.93753253992


In [18]:
# медиана по ежемесячному доходу, основываясь на типе занятости

grp = df.groupby(['income_type'])
df['total_income'] = grp.total_income.apply(lambda x: x.fillna(x.median()))

df.info()
df['total_income'].describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


count    2.152500e+04
mean     1.652253e+05
std      9.804367e+04
min      2.066726e+04
25%      1.077982e+05
50%      1.425944e+05
75%      1.955499e+05
max      2.265604e+06
Name: total_income, dtype: float64

<div class="alert alert-block alert-danger">
<s><b>Комментарий от ревьюера</b>

Замена на обычную медиану даст нам искаженные данные. Смотри,у нас много разных профессий. Давай заменим пропущенные значения на медиану по профессии.
    
    
</div>



Пропуски успешно были заполнены.

<div class="alert alert-block alert-info">
<b>Комментарий от студента</b> <a class="tocSkip"></a>
    
Далее данные были изменены, проверил выборочно группы людей, дабы не нагружать исследование не стал включать доп проверки, а решил оставить комментарий. Таким образом н-р значение у пенсионеров раньше = "145017.9" (что соответствует медиане по столбцу), новое "118514.4". Но пришлось фиксить предупреждение. Однако, считаю этот метод лаконичнее функции, и позволяет использовать встроенные методы pandas :)
</div>

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера. V.2.</b>

Еще проверки это хорошо и на самом деле было бы правильно их включить. Чтобы точно быть во всем уверенным. :)

Warning(предупреждение) можно просто скрыть,не обязательно его слушать. :)</div>    

#### Шаг 2.2 Исправление анамалий

In [19]:
# Удалим из df строки с аномалиями в "Children" и -1 заменим на 1
df['children'] = df['children'].replace(-1, 1)
df = df.loc[df['children'] != 20]

# dob_years": заменить 0 на медианное значение.

dob_years_median = df.loc[df.loc[:, 'dob_years'] != 0]['dob_years'].median()
df['dob_years'] = df['dob_years'].replace(0, dob_years_median)

# Проверим

df.info()
display(df['children'].describe())
display(df['dob_years'].describe())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21449 entries, 0 to 21524
Data columns (total 12 columns):
children            21449 non-null int64
days_employed       19284 non-null float64
dob_years           21449 non-null int64
education           21449 non-null object
education_id        21449 non-null int64
family_status       21449 non-null object
family_status_id    21449 non-null int64
gender              21449 non-null object
income_type         21449 non-null object
debt                21449 non-null int64
total_income        21449 non-null float64
purpose             21449 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


count    21449.000000
mean         0.474334
std          0.751417
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max          5.000000
Name: children, dtype: float64

count    21449.000000
mean        43.499091
std         12.222229
min         19.000000
25%         34.000000
50%         43.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

Успешно удалили и заменили значения.

In [20]:
# Удалим столбец "days_employed"
df.drop(columns=['days_employed'], axis=1, inplace=True)

# Столбец "gender": сменить гендр "XNA"
df.loc[df['gender'] == 'XNA', 'gender'] = 'F'

# Проверим
print(df['gender'].value_counts())
print()
df.info()

F    14190
M     7259
Name: gender, dtype: int64

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21449 entries, 0 to 21524
Data columns (total 11 columns):
children            21449 non-null int64
dob_years           21449 non-null int64
education           21449 non-null object
education_id        21449 non-null int64
family_status       21449 non-null object
family_status_id    21449 non-null int64
gender              21449 non-null object
income_type         21449 non-null object
debt                21449 non-null int64
total_income        21449 non-null float64
purpose             21449 non-null object
dtypes: float64(1), int64(5), object(5)
memory usage: 2.0+ MB


Успешно обновили данные.

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера</b>

Всё отлично! :)
    
</div>

#### Шаг 2.3. Изменение типов данных.

In [21]:
df['total_income'] = df['total_income'].astype(int)

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21449 entries, 0 to 21524
Data columns (total 11 columns):
children            21449 non-null int64
dob_years           21449 non-null int64
education           21449 non-null object
education_id        21449 non-null int64
family_status       21449 non-null object
family_status_id    21449 non-null int64
gender              21449 non-null object
income_type         21449 non-null object
debt                21449 non-null int64
total_income        21449 non-null int64
purpose             21449 non-null object
dtypes: int64(6), object(5)
memory usage: 2.0+ MB


Сменили тип данных для удобства работы.

#### Шаг 2.4. Удаление дубликатов.

In [22]:
# Столбец "education": привести в общий регистр.
df['education'] = df['education'].str.lower()

print(df['education'].value_counts())
print()
print('Всего явных дубликатов', df.duplicated().sum())

среднее                15173
высшее                  5246
неоконченное высшее      742
начальное                282
ученая степень             6
Name: education, dtype: int64

Всего явных дубликатов 71


71 явный дубликат, избавимся от них.

In [23]:
df = df.drop_duplicates().reset_index(drop=True)

print('Всего явных дубликатов', df.duplicated().sum())
print()
df.info()

Всего явных дубликатов 0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21378 entries, 0 to 21377
Data columns (total 11 columns):
children            21378 non-null int64
dob_years           21378 non-null int64
education           21378 non-null object
education_id        21378 non-null int64
family_status       21378 non-null object
family_status_id    21378 non-null int64
gender              21378 non-null object
income_type         21378 non-null object
debt                21378 non-null int64
total_income        21378 non-null int64
purpose             21378 non-null object
dtypes: int64(6), object(5)
memory usage: 1.8+ MB


Данные успешно обновлены.

#### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [24]:
# создадим словари

education_dict = df[['education_id', 'education', 'debt']]
family_dict = df[['family_status_id', 'family_status', 'debt']]

df.drop(columns=['education', 'family_status'], axis=1, inplace=True)
df.info()
df.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21378 entries, 0 to 21377
Data columns (total 9 columns):
children            21378 non-null int64
dob_years           21378 non-null int64
education_id        21378 non-null int64
family_status_id    21378 non-null int64
gender              21378 non-null object
income_type         21378 non-null object
debt                21378 non-null int64
total_income        21378 non-null int64
purpose             21378 non-null object
dtypes: int64(6), object(3)
memory usage: 1.5+ MB


children  dob_years  education_id  family_status_id gender income_type  \
0         1         42             0                 0      F   сотрудник   
1         1         36             1                 0      F   сотрудник   
2         0         33             1                 0      M   сотрудник   
3         3         32             1                 0      M   сотрудник   
4         0         53             1                 1      F   пенсионер   
5         0         27             0                 1      M   компаньон   
6         0         43             0                 0      F   компаньон   
7         0         50             1                 0      M   сотрудник   
8         2         35             0                 1      F   сотрудник   
9         0         41             1                 0      M   сотрудник   

   debt  total_income                     purpose  
0     0        253875               покупка жилья  
1     0        112080     приобретение автомобиля  
2     0        145885               покупка жилья  
3     0        267628  дополнительное образование  
4     0        158616             сыграть свадьбу  
5     0        255763               покупка жилья  
6     0        240525           операции с жильем  
7     0        135823                 образование  
8     0         95856       на проведение свадьбы  
9     0        144425     покупка жилья для семьи

Преобразования прошли успешно.

#### Шаг 2.6. Категоризация дохода.

In [25]:
# Ошибки записи и некорректные строки.
wrong_lines_content = []
wrong_lines = 0

# Напишем функцию категоризации.
def income_category(row):
    try:
        if row <= 30000:
            return '"E" до 30000'
        if row <= 50000:
            return '"D" до 50000'
        if row <= 200000:
            return '"C" до 200000'
        if row <= 1000000:
            return '"B" до 1000000'
        return '"A" более 1000000'
    except:
        wrong_lines += 1
        wrong_lines_content.append(row)


# Добавим её в столбец.
df['total_income_category'] = df['total_income'].apply(income_category)

# Проверим некорректные строки.
print('Количество некорректных строк', wrong_lines)
print('Некорректные строки', wrong_lines_content)  

# Проверим таблицу.
display(df.head(10))
print()
df.info()

Количество некорректных строк 0
Некорректные строки []


children  dob_years  education_id  family_status_id gender income_type  \
0         1         42             0                 0      F   сотрудник   
1         1         36             1                 0      F   сотрудник   
2         0         33             1                 0      M   сотрудник   
3         3         32             1                 0      M   сотрудник   
4         0         53             1                 1      F   пенсионер   
5         0         27             0                 1      M   компаньон   
6         0         43             0                 0      F   компаньон   
7         0         50             1                 0      M   сотрудник   
8         2         35             0                 1      F   сотрудник   
9         0         41             1                 0      M   сотрудник   

   debt  total_income                     purpose total_income_category  
0     0        253875               покупка жилья        "B" до 1000000  
1     0        112080     приобретение автомобиля         "C" до 200000  
2     0        145885               покупка жилья         "C" до 200000  
3     0        267628  дополнительное образование        "B" до 1000000  
4     0        158616             сыграть свадьбу         "C" до 200000  
5     0        255763               покупка жилья        "B" до 1000000  
6     0        240525           операции с жильем        "B" до 1000000  
7     0        135823                 образование         "C" до 200000  
8     0         95856       на проведение свадьбы         "C" до 200000  
9     0        144425     покупка жилья для семьи         "C" до 200000


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21378 entries, 0 to 21377
Data columns (total 10 columns):
children                 21378 non-null int64
dob_years                21378 non-null int64
education_id             21378 non-null int64
family_status_id         21378 non-null int64
gender                   21378 non-null object
income_type              21378 non-null object
debt                     21378 non-null int64
total_income             21378 non-null int64
purpose                  21378 non-null object
total_income_category    21378 non-null object
dtypes: int64(6), object(4)
memory usage: 1.6+ MB


Успешно применили функцию и создали столбец, все строки записались без проблем.

<div class="alert alert-block alert-warning">
<b>Комментарий от ревьюера</b>

Что значит А,В,С...?  Если не будет описания, мы никогда не поймем это. Лучше добавлять конкретные цифры или сделать словарь.
    
</div>

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера. V.2.</b>

О,теперь понятно. Тут еще есть вариант создать словарь. :)</div>    

#### Шаг 2.7. Категоризация целей кредита.

In [26]:
# Ошибки записи и некорректные строки.
wrong_lines_content = []
wrong_lines = 0


# Напишем функцию категоризации целей кредита.
def purpose_f(row):
    try:
        if 'образовани' in row:
            return 'получение образования'
        if 'автомобил' in row:
            return 'операции с автомобилем'
        if 'свадьб' in row:
            return 'проведение свадьбы'
        return 'операции с недвижимостью'
    except:
        wrong_lines += 1
        wrong_lines_content.append(row)


# Добавим столбец.
df['purpose_category'] = df['purpose'].apply(purpose_f)

# Проверим некорректные строки.
print('Количество некорректных строк', wrong_lines)
print('Некорректные строки', wrong_lines_content)

# Проверим датафрейм.
display(df.head(10))
df.info()

Количество некорректных строк 0
Некорректные строки []


children  dob_years  education_id  family_status_id gender income_type  \
0         1         42             0                 0      F   сотрудник   
1         1         36             1                 0      F   сотрудник   
2         0         33             1                 0      M   сотрудник   
3         3         32             1                 0      M   сотрудник   
4         0         53             1                 1      F   пенсионер   
5         0         27             0                 1      M   компаньон   
6         0         43             0                 0      F   компаньон   
7         0         50             1                 0      M   сотрудник   
8         2         35             0                 1      F   сотрудник   
9         0         41             1                 0      M   сотрудник   

   debt  total_income                     purpose total_income_category  \
0     0        253875               покупка жилья        "B" до 1000000   
1     0        112080     приобретение автомобиля         "C" до 200000   
2     0        145885               покупка жилья         "C" до 200000   
3     0        267628  дополнительное образование        "B" до 1000000   
4     0        158616             сыграть свадьбу         "C" до 200000   
5     0        255763               покупка жилья        "B" до 1000000   
6     0        240525           операции с жильем        "B" до 1000000   
7     0        135823                 образование         "C" до 200000   
8     0         95856       на проведение свадьбы         "C" до 200000   
9     0        144425     покупка жилья для семьи         "C" до 200000   

           purpose_category  
0  операции с недвижимостью  
1    операции с автомобилем  
2  операции с недвижимостью  
3     получение образования  
4        проведение свадьбы  
5  операции с недвижимостью  
6  операции с недвижимостью  
7     получение образования  
8        проведение свадьбы  
9  операции с недвижимостью

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21378 entries, 0 to 21377
Data columns (total 11 columns):
children                 21378 non-null int64
dob_years                21378 non-null int64
education_id             21378 non-null int64
family_status_id         21378 non-null int64
gender                   21378 non-null object
income_type              21378 non-null object
debt                     21378 non-null int64
total_income             21378 non-null int64
purpose                  21378 non-null object
total_income_category    21378 non-null object
purpose_category         21378 non-null object
dtypes: int64(6), object(5)
memory usage: 1.8+ MB


<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера</b>

Всё прошло успешно! Ура! :)    
</div>

### Ответы на вопросы.

##### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

Создадим таблицу, где будем обозревать данные

In [27]:
df_pivot = df.pivot_table(index=['gender'], columns='children', values='debt', aggfunc='sum')
df_pivot

children    0    1    2   3  4  5
gender                           
F         592  246  134  17  1  0
M         471  199   60  10  3  0

По таблице уже видно, что прослеживается взаимосвязь, для большей наглядности посчитаем % тех, кто не оплатил задолженность вовремя.

In [28]:
# Функция принимает список столбцов, или столбец и выводит таблицу
def a_debt(column_list):
    a_table = df.groupby(column_list).agg({'debt':['sum', 'count']})
    a_table[('debt', '%')] = a_table[('debt', 'sum')] / a_table[('debt', 'count')]*100
    return a_table

a_debt(['children'])

debt                 
           sum  count         %
children                       
0         1063  14091  7.543822
1          445   4855  9.165808
2          194   2052  9.454191
3           27    330  8.181818
4            4     41  9.756098
5            0      9  0.000000

##### Вывод:

Есть небольшая взаимосвязь между количеством детей и выплатами, меньше всего просрочки у тех, у кого детей вовсе нет и многодетных семей с 3 детьми, больше всего просрочек у семей с 1-2 детьми и семей с 4.

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера</b>

5 детей исключение точно. Ты прав, дети не способствуют выплате кредита вовремя. :) 
</div>

#### Есть ли зависимость между семейным положением и возвратом кредита в срок?

Составим таблицу, где будет наглядно отображено:

- статус семьи.
- посчитаем должников.
- посчитакм % соотношение должников к общей сумме.

In [29]:
pi_fam_stat = family_dict.groupby(['family_status_id', 'family_status']).agg({'debt':['sum', 'count']})
pi_fam_stat[('debt','%')] = pi_fam_stat[('debt', 'sum')] / pi_fam_stat[('debt', 'count')]*100

pi_fam_stat

debt                 
                                        sum  count         %
family_status_id family_status                              
0                женат / замужем        928  12290  7.550854
1                гражданский брак       385   4139  9.301764
2                вдовец / вдова          63    955  6.596859
3                в разводе               84   1193  7.041073
4                Не женат / не замужем  273   2801  9.746519

##### Вывод:

Исходя из данных, можно сделать вывод, что люди которые живут не в браке, и люди которые живут в гражданском браке имеют низкий показатель выплаты кредитов в срок. Женатых людей можно отнести к средней категории доверенности. Вовремя кредиты как правило платят люди в разводе, и те кто овдовели.

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера</b>

Да, всё так. </div>

#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Воспользуемся функцией, которую составили выше:

Также, уровни зароботка: 
- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

In [30]:
a_debt(['total_income_category'])

debt                 
                        sum  count         %
total_income_category                       
"A" более 1000000         2     25  8.000000
"B" до 1000000          354   5022  7.048984
"C" до 200000          1354  15959  8.484241
"D" до 50000             21    350  6.000000
"E" до 30000              2     22  9.090909

##### Вывод:

Самый низкий уровень просрочки у людей с достатком категории: 'D', и 'B', средний уровень доверенности заслуживают категория: 'А'. Как ни странно, но просрочка по кредитам чаще встречается у людей либо самого низкого достатка, категории 'E', либо среднего "С"

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера</b>


Наши выводы здесь были ожидаемы. :) </div>

<div class="alert alert-block alert-info">
<b>Комментарий от студента</b> <a class="tocSkip"></a>
    
Тут повлияли на данные наши перерассчёты :)
</div>

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера. V.2.</b>

Тоже вполне себе ожидаемо. Вот если бы у самых богатых были долги, а бедных наоборот, то был бы сюрприз.</div>    

#### Как разные цели кредита влияют на его возврат в срок?

Также как раннее, воспользуемся функцией:

In [31]:
a_debt(['purpose_category'])

debt                 
                          sum  count         %
purpose_category                              
операции с автомобилем    401   4290  9.347319
операции с недвижимостью  780  10775  7.238979
получение образования     369   3998  9.229615
проведение свадьбы        183   2315  7.904968

Можно заметить, что кредиты связанные с недвижимостью и проведением свадьбы чаще всего отдают в срок. Больше всего просрочек с целью кредита: авто и образование.

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера</b>

При кредите на недвижимость столько проверок, что неудивительно, что они имею меньше должников. :)</div>

## Общий вывод:

В ходе предобработки мы удалили столбец "days_employed", так как большинство данных не соответствовали реальности, и восстанавливать данные для поставленных задач смысла не имело. Отработали и удалили аномалии/дубликаты в столбцах (Аномальное кол-во детей, и 0 лет человеку), создали словари для удобства работы с таблицей и объединили цели кредитов.

__Основные выводы исследования:__
- Есть прямая зависимость, между количеством детей и выплате кредита вовремя, чем меньше детей, тем более исправно люди платят кредит
- Те пары, которые состоят в гражданском браке, либо холосты, имеют больше всех просрочек по выплатам. Женатые, вдовцы, и в разводе: охотнее платят кредиты вовремя.
- Люди с низким заработком (до 30 тыс.) и люди с средним достатком (до 200 тыс.) имеют больший процент просрочки. Вовремя же платят люди с высоким (от 200 тыс. до 1 млн.) и средне-низким (от 30 тыс. до 50 тыс.)
- Цели кредита влияют на его выплату в срок: так те кто берёт недвижимость или проводят свадьбы платят кредит вовремя, а те кто покупает автомобиль, либо оформляют кредит на учёбу, в основном имеют большую просрочку.

<div class="alert alert-block alert-danger">
<s><b>Комментарий от ревьюера</b>

Финальный вывод - это чаще всего единственное, что читает клиент/менеджер.  Тут нужно описать всё: что у нас было, что добавили/поменяли,почему, что получилось с нашими вопросами?</div>

<div class="alert alert-block alert-info">
<b>Комментарий от студента</b> <a class="tocSkip"></a>
    
Оля, спасибо за твою работу, и чудесного дня :)
</div>

<div class="alert alert-block alert-success">
<b>Комментарий от ревьюера. V.2.</b>

Отличный вывод получился! Спасибо за проект! :) </div>    